cmd/event-notification.go

```diff
@@ -31,23 +32,39 @@ import (
 	"github.com/minio/pkg/bucket/policy"
 )
 
+const (
+	// Values for bootstrap level
+	eventNotifierInitializing     uint32 = iota + 1 // initialized but without targets
+	eventNotifierFullyInitialized                   // initialized with targets
+)
+
 // EventNotifier - notifies external systems about events in MinIO.
 type EventNotifier struct {
-	sync.RWMutex
+	// Store events in a queue since the object layer can be ready
+	// before the config subsystem is ready and do not have yet
+	// the list of the configured notification targets
+	bootstrapLevel   uint32
+	bootQueue        chan eventArgs
+	bootMu           sync.Mutex
+	bootQueueClosing bool
```

Comment:
Why not have the queue on permanently instead of "turning it off" after init? That way, we won't need the `bootMu` lock and the closing flag. It will also limit the number of threads (currently equal to number of requests sending notifications) that could be blocked sending out the notifications to external systems.